### DISCUSSION TITLE
language server: suggest functions based on first parameter

### DISCUSSION THREAD

**Discussion by orvitpng** [gleam-lang/gleam#3366] [Ideas & suggestions]
For the purposes of this suggestion, we will use the following code:
```gleam
let value = "Hello, world!"
```

Now, take into account imported modules and their functions, filter out all functions where the first parameter is a String, and then provide that to the client. For instance, an end user could do `value.leng [Enter]` and that would convert to `string.length(value)`.

**Comment by lpil** [comment]
I presume that there should be a |> in the example code there?

What would the programmer do when they want a function where the first argument is not of that type? One can pipe to any position in Gleam.

**Reply by Shad0wRim** [reply]
I believe the intention is to allow the language server to autocomplete OOP style method access into passing into the function directly.
```gleam
let value = "Hello, world!"
value.leng
```
Typing `value.leng [Enter]` would have the LS autocomplete into `string.length(value)`

**Reply by lpil** [reply]
Right, which is why I asked the question above. Gleam does not have the same limitations as OOP here, the language permits you to be much more flexible with what functions you "chain".

**Reply by Shad0wRim** [reply]
I would include functions with the type in any parameter position, but sort them by order of the parameter, first argument at the top, second lower, etc., however I am not familiar with LSP development so I am not sure if this is feasible.

I personally would prefer this with only the pipe operator rather than dot notation, so that chaining feels more natural, but perhaps this style could be used for single arity functions, expanding into a pass into the function rather than a pipe, like a derived property access in Python.
